File: PracticePlatform/Scheduling/backend/app/services/patient_service.py
"""
Patient Service Integration
Handles API communication with the Patient microservice from HealthcareFoundation/CoreServices
"""

import httpx
import asyncio
from typing import List, Optional, Dict, Any
from fastapi import HTTPException
import os

class PatientService:
    def __init__(self):
        self.base_url = os.getenv("PATIENT_SERVICE_URL", "http://patient-service:8000")
        self.timeout = 30.0
    
    async def _make_request(self, method: str, endpoint: str, **kwargs) -> Dict[str, Any]:
        """Make HTTP request to patient service"""
        url = f"{self.base_url}{endpoint}"
        
        try:
            async with httpx.AsyncClient(timeout=self.timeout) as client:
                response = await client.request(method, url, **kwargs)
                response.raise_for_status()
                return response.json()
        except httpx.HTTPStatusError as e:
            if e.response.status_code == 404:
                raise HTTPException(status_code=404, detail="Patient not found")
            elif e.response.status_code == 400:
                raise HTTPException(status_code=400, detail="Invalid request to patient service")
            else:
                raise HTTPException(status_code=500, detail=f"Patient service error: {e.response.text}")
        except httpx.RequestError as e:
            raise HTTPException(status_code=503, detail=f"Patient service unavailable: {str(e)}")
    
    async def get_patients(self, skip: int = 0, limit: int = 10, active: Optional[bool] = None) -> List[Dict[str, Any]]:
        """Get list of patients"""
        params = {"skip": skip, "limit": limit}
        if active is not None:
            params["active"] = active
        
        return await self._make_request("GET", "/patients", params=params)
    
    async def get_patient_by_id(self, patient_id: str) -> Dict[str, Any]:
        """Get patient by UUID"""
        return await self._make_request("GET", f"/patients/{patient_id}")
    
    async def get_patient_by_fhir_id(self, fhir_id: str) -> Dict[str, Any]:
        """Get patient by FHIR ID"""
        return await self._make_request("GET", f"/patients/fhir/{fhir_id}")
    
    async def create_patient(self, patient_data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new patient"""
        return await self._make_request("POST", "/patients", json=patient_data)
    
    async def update_patient(self, patient_id: str, patient_data: Dict[str, Any]) -> Dict[str, Any]:
        """Update an existing patient"""
        return await self._make_request("PUT", f"/patients/{patient_id}", json=patient_data)
    
    async def delete_patient(self, patient_id: str) -> Dict[str, Any]:
        """Delete a patient"""
        return await self._make_request("DELETE", f"/patients/{patient_id}")
    
    async def search_patients_by_name(self, family_name: Optional[str] = None, given_name: Optional[str] = None, limit: int = 10) -> List[Dict[str, Any]]:
        """Search patients by name"""
        params = {"limit": limit}
        if family_name:
            params["family_name"] = family_name
        if given_name:
            params["given_name"] = given_name
        
        return await self._make_request("GET", "/patients/search/name", params=params)
    
    async def health_check(self) -> Dict[str, Any]:
        """Check patient service health"""
        return await self._make_request("GET", "/health")

# Global instance
patient_service = PatientService() 

File: PracticePlatform/Scheduling/frontend/src/services/patientService.js
import axios from 'axios';

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:8000';

// Create axios instance with default config
const apiClient = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Add request interceptor to include auth token
apiClient.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('token');
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

export const patientService = {
  // Get all patients with pagination
  getPatients: async (params = {}) => {
    try {
      const response = await apiClient.get('/patients/', { params });
      return response.data;
    } catch (error) {
      console.error('Error fetching patients:', error);
      throw error;
    }
  },

  // Get patient by ID
  getPatientById: async (patientId) => {
    try {
      const response = await apiClient.get(`/patients/${patientId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching patient:', error);
      throw error;
    }
  },

  // Get patient by FHIR ID
  getPatientByFhirId: async (fhirId) => {
    try {
      const response = await apiClient.get(`/patients/fhir/${fhirId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching patient by FHIR ID:', error);
      throw error;
    }
  },

  // Create new patient
  createPatient: async (patientData) => {
    try {
      const response = await apiClient.post('/patients/', patientData);
      return response.data;
    } catch (error) {
      console.error('Error creating patient:', error);
      throw error;
    }
  },

  // Update patient
  updatePatient: async (patientId, patientData) => {
    try {
      const response = await apiClient.put(`/patients/${patientId}`, patientData);
      return response.data;
    } catch (error) {
      console.error('Error updating patient:', error);
      throw error;
    }
  },

  // Delete patient
  deletePatient: async (patientId) => {
    try {
      const response = await apiClient.delete(`/patients/${patientId}`);
      return response.data;
    } catch (error) {
      console.error('Error deleting patient:', error);
      throw error;
    }
  },

  // Search patients by name
  searchPatientsByName: async (searchParams) => {
    try {
      const response = await apiClient.get('/patients/search/name', { params: searchParams });
      return response.data;
    } catch (error) {
      console.error('Error searching patients:', error);
      throw error;
    }
  },

  // Search patients (POST method)
  searchPatients: async (searchData) => {
    try {
      const response = await apiClient.post('/patients/search', searchData);
      return response.data;
    } catch (error) {
      console.error('Error searching patients:', error);
      throw error;
    }
  },

  // Health check
  healthCheck: async () => {
    try {
      const response = await apiClient.get('/patients/health/check');
      return response.data;
    } catch (error) {
      console.error('Error checking patient service health:', error);
      throw error;
    }
  }
};

export default patientService; 